### ISSUE TITLE
System.NullReferenceException thrown when accessing UserAsync.GetAuthenticatedUser()

### ISSUE THREAD

**Issue by jdahan91** [linvi/tweetinvi#850]
Yes, I have set the credentials flag by calling the Auth.SetCredentials() but I have no clue why the reason for this exception.

**Comment by GodsWithin** [comment]
I'm having the same issue right now. It looks like Json.NET also released a new version, which might be partially responsible for it?

**Comment by jdahan91** [comment]
> Edit: this is the same case when using User.GetAuthenticatedUser()

Yes.

I think linvi needs to release an emergency update.

**Comment by jdahan91** [comment]
> It looks like Json.NET also released a new version, which might be partially responsible for it?

Could be even a change in twitter's api.

**Comment by DeschaumesDaan** [comment]
I'm facing the similar issue, can't get any data from the API, though it worked perfectly two hours ago. 
I know Twitter is busy rebuilding their APIs, perhaps that's the issue?

**Comment by GodsWithin** [comment]
https://github.com/linvi/tweetinvi/blob/master/Tweetinvi.Logic/JsonConverters/JsonLanguageConverter.cs#L14 on this line, the reader.Value is null, and when you call reader.Value.ToString() it'll throw the NullReferenceException, which causes everything to break.

So it seems that it has to do with the Json.NET library.

**Comment by GodsWithin** [comment]
> https://github.com/linvi/tweetinvi/blob/master/Tweetinvi.Logic/JsonConverters/JsonLanguageConverter.cs#L14 on this line, the reader.Value is null, and when you call reader.Value.ToString() it'll throw the NullReferenceException, which causes everything to break.
> 
> So it seems that it has to do with the Json.NET library.

If you add a null check before trying to parse, it'll not throw the exception and from there no other exceptions will be thrown and the GetAuthenticatedUser() method will return properly.

**Comment by DeschaumesDaan** [comment]
@GodsWithin Thanks, I will try this.

**Comment by jdahan91** [comment]
@GodsWithin Can you make a pull request? I need to release a new version tonight.

**Comment by GodsWithin** [comment]
@jdahan91 Done

**Comment by jcmag** [comment]
> 
> 
> > https://github.com/linvi/tweetinvi/blob/master/Tweetinvi.Logic/JsonConverters/JsonLanguageConverter.cs#L14 on this line, the reader.Value is null, and when you call reader.Value.ToString() it'll throw the NullReferenceException, which causes everything to break.
> > So it seems that it has to do with the Json.NET library.
> 
> If you add a null check before trying to parse, it'll not throw the exception and from there no other exceptions will be thrown and the GetAuthenticatedUser() method will return properly.

I've just applied this fix and it works again

**Comment by GodsWithin** [comment]
> > > https://github.com/linvi/tweetinvi/blob/master/Tweetinvi.Logic/JsonConverters/JsonLanguageConverter.cs#L14 on this line, the reader.Value is null, and when you call reader.Value.ToString() it'll throw the NullReferenceException, which causes everything to break.
> > > So it seems that it has to do with the Json.NET library.
> > 
> > 
> > If you add a null check before trying to parse, it'll not throw the exception and from there no other exceptions will be thrown and the GetAuthenticatedUser() method will return properly.
> 
> I've just applied this fix and it works again

As I mentioned in the PR, I don't think this is a proper fix since we don't know where the problem originates from (being the Value property of the JsonReader that is null).

**Comment by z0o0x** [comment]
I have the same problem here please we need a fix for it
I have +200 users the app stop working for him

**Comment by jdahan91** [comment]
Guys, linvi (chatting with hom via Skype) told me he's busy with work and don't have time to release a hotfix. He said our current option is to download the lib source code, fix and compile it until 5.0 is released which is 1-2 month away. @GodsWithin he said your fix is not good enough, even when I told him that it's temporary lol.

**Comment by GodsWithin** [comment]
@jdahan91 Then I suggest you guys just do as linvi said, download the code, open in Visual Studio, add that piece of code to hotfix it, build and use the resulting .dll to replace the one you're currently using.

**Comment by jdahan91** [comment]
Ok, anyone else having issues with the stream api? event handlers are not being fired.. Anyone can issue a temp fix?

**Comment by GodsWithin** [comment]
@jdahan91 just noticed that the stream stopped working for me as well. I've no idea how that code works though.

**Comment by jdahan91** [comment]
This library is not in sync with Twitter's api.

**Comment by GodsWithin** [comment]
Did Twitter change something the last few hours? All of a sudden?

**Comment by GodsWithin** [comment]
@jdahan91 I think this might be useful https://twittercommunity.com/t/upcoming-changes-to-user-object-and-get-users-suggestions-endpoints/124732

**Comment by GodsWithin** [comment]
@jdahan91 At the moment my FilteredStream seems to be working fine.

**Comment by jdahan91** [comment]
> @jdahan91 I think this might be useful https://twittercommunity.com/t/upcoming-changes-to-user-object-and-get-users-suggestions-endpoints/124732

Indeed this is useful. Especially for the lead developer, linvi.

**Comment by jdahan91** [comment]
> @jdahan91 At the moment my FilteredStream seems to be working fine.

Working for me too, but with ti y delay and the handler doesn't get fired all the time.

**Comment by z0o0x** [comment]
guys do you have any steps i can do it to fix this ?
upload dll working fine 
now i have to update the app on google play

**Comment by GodsWithin** [comment]
@z0o0x This is a hotfix for the NullReferenceException https://github.com/linvi/tweetinvi/pull/852, you'll have to clone, change the line and build the .dll yourself.

**Comment by z0o0x** [comment]
@GodsWithin  I don't know how to do it
can you upload your dll so i can use it ?

**Comment by GodsWithin** [comment]
@z0o0x https://github.com/GodsWithin/tweetinvi/blob/master/Tweetinvi.Logic.dll
For reference, how you'd build this yourself. You open the solution in Visual Studio, change the code.

You open a Command Prompt and `cd` into the same directory as the solution. Next you use the command: `dotnet publish TweeiInvi.Logic -r <RID>`, in my case was my Runtime ID win10-x64 (Windows 10 64-bit). This will build that project and the .dll will be in `...\Tweetinvi.Logic\bin\Debug\netstandard1.4`

**Comment by EwanNoble** [comment]
My filtered stream has broken today with a NullReferenceException, unfortunately your fix doesn't resolve it ☹️

**Comment by z0o0x** [comment]
@GodsWithin  I tried for 3 hour to add this reference to project 

https://github.com/GodsWithin/tweetinvi/blob/master/Tweetinvi.Logic.dll

But i don't know how to do it i am a beginner sorry xD

I just need it to work 
If you can help me and the others how to fix this step by step i will be very thankful
Just let us know how to add reference of your dll and make tweetinvi working again 

Case "fix and compile it until 5.0 is released which is 1-2 month away" i can't wait for 2 month to fix problem with update :(

**Comment by soatley** [comment]
In my case, I downloaded the ZIP file and just saved it to my hard drive.  Opened the Tweetinvi.sln file in Visual Studio, made the correction to the file/line that @GodsWithin  suggested (https://github.com/linvi/tweetinvi/issues/850#issuecomment-494501533) and then built the solution.  

In my main solution I use TweetInvi in, I removed the NuGet version of TweetInvi and then added the DLL files from the TweetInvi solution I just finished building.  Test and release.  Worked great.

When something works so well for so long without a hiccup, we get spoiled.  Thanks to everyone for the quick reaction and suggestions.

**Comment by z0o0x** [comment]
Hi again

Thank you All
i fix the issue after I downloaded the ZIP like what @soatley suggested 
then i do what  @GodsWithin suggested (#850 (comment)) then built the solution.

after that in my main, I removed the NuGet version of TweetInvi and then added the DLL files from the TweetInvi solution from path all dll
....\tweetinvi-master\Tweetinvi\bin\Debug\netstandard1.4
and 
Nito.AsyncEx.Context.dll
Nito.AsyncEx.Tasks.dll
autofac.dll

and working for me great Thank You all guys <3

**Comment by jxlarrea** [comment]
Just wanted to confirm that building Tweetinvi from source with #852 patch fixes the issue.  Thanks guys for the help.

**Comment by atobtech** [comment]
Cloned the code, applied above mentioned temp fix, rebuild, referenced updated dll. Works fine now but streamer is now slower than before

**Comment by eranruso** [comment]
> @GodsWithin I tried for 3 hour to add this reference to project
> 
> https://github.com/GodsWithin/tweetinvi/blob/master/Tweetinvi.Logic.dll
> 
> But i don't know how to do it i am a beginner sorry xD
> 
> I just need it to work
> If you can help me and the others how to fix this step by step i will be very thankful
> Just let us know how to add reference of your dll and make tweetinvi working again
> 
> Case "fix and compile it until 5.0 is released which is 1-2 month away" i can't wait for 2 month to fix problem with update :(

The easiest way is to go to your project's references, see that the dll is referenced already.
Remove it from the references and add a new reference, than choose browse and go to the new dll, choose it and you are done!
make sure you have the new dll on the disk before starting this process

**Comment by jdahan91** [comment]
> Cloned the code, applied above mentioned temp fix, rebuild, referenced updated dll. Works fine now but streamer is now slower than before

Yes, indeed the stream is slow. Original discussion should take place here #851

**Comment by LambertWM** [comment]
I don't know how to use GitHub properly, but my suggestion would be to temporarily "fix" the problem like in the below code fragment.

After all ```LanguageExtension.GetLangFromDescription( ```) returns a ```Language``` enum value so returning ```Language.English``` in case of trouble makes more sense to me than the earlier suggestion, which returns a ```bool``` expression wrapped in an ```object``` (?)

```
public override object ReadJson(JsonReader reader, Type objectType, object existingValue, JsonSerializer serializer)
{
	int parsed;

	// LWM: TEMPFIX: if we do not have a reader or if we have a reader with a null Value, then return a sensible default value
	if(reader?.Value==null)
		return Language.English;

	if (int.TryParse(reader.Value.ToString(), out parsed))
	{
		return LanguageExtension.GetLangFromDescription(parsed);
	}

	return LanguageExtension.GetLangFromDescription((string)reader.Value);
}
```

**Comment by dynamicons** [comment]
Tried out the solution proposed, set the TweetInvi.Logic.dll version to 4.0.0.1, recompiled.

Opened my project.
in Project Explorer, show Library References, exclude Tweetinvi.Logic.dll, add your custom one.

Recompiled, works.
Thank you for supporting us.

Martin Bernhardt, Founder
www.dynamic-applications.com

**Comment by rooonikz** [comment]
Hi guys. Just faced this issue and found some extension points not to execute local clone - rebuild steps. Just use `JsonPropertyConverterRepository` like in example below. Works for me.
`JsonPropertyConverterRepository.JsonConverters.Remove(typeof(Language));
            JsonPropertyConverterRepository.JsonConverters.Add(typeof(Language), new CustomJsonLanguageConverter()); // your brand new json converter with workaround`

`    public class CustomJsonLanguageConverter : JsonLanguageConverter
    {
        public override object ReadJson(JsonReader reader, Type objectType, object existingValue, JsonSerializer serializer)
        {
            return reader.Value != null 
                ? base.ReadJson(reader, objectType, existingValue, serializer) 
                : Language.English;
        }
    }`

Hope that will help somebody.

**Comment by Joakim432710** [comment]
Guys, https://developer.twitter.com/en/docs/tweets/data-dictionary/overview/user-object.html 
Language is deprecated for the IUser object, 
"lang | null | Value will be set to null. Still available via GET account/settings as language"

PR: https://github.com/linvi/tweetinvi/pull/856

**Comment by steve-lonie** [comment]
I had this issue today with a really old version of the project. I updated the nuget packages and used patched TweetInvi.Logic.dll and everything is back to normal.

**Comment by GodsWithin** [comment]
People using my Tweetinvi.Logic.dll, please note that it was a very quick, low effort, not much thought put into it patch.

I believe the solutions provided after are "cleaner".
https://github.com/linvi/tweetinvi/issues/850#issuecomment-494743288
https://github.com/linvi/tweetinvi/issues/850#issuecomment-494746515

Not sure if https://github.com/linvi/tweetinvi/pull/856 would break anything else though.

**Comment by z0o0x** [comment]
Guy i believe these solutions are temporary and works good
Until tweetinvi 5 soon

**Comment by JaneSmith** [comment]
I think I'm running into this issue maybe. I've had an online service that's been working for a year, and today it suddenly stopped working. When I call `Timeline.GetUserTimeline`, I get a `NullReferenceException`, and the stack trace says it occurred at `Tweetinvi.Logic.JsonConverters.JsonLanguageConverter.ReadJson`...

**Comment by jdahan91** [comment]
@rooonikz Please share a modified .dll according to your solution.

**Comment by miteshkwan** [comment]
How can we compile the code for .net core?

**Comment by miteshkwan** [comment]
@rooonikz can you explain it step by step for newbies?

**Comment by rooonikz** [comment]
> @rooonikz Please share a modified .dll according to your solution.

I didn't modify original Nuget package, just added mentioned code somewhere in my codebase.

**Comment by rooonikz** [comment]
> @rooonikz can you explain it step by step for newbies?

Just create new class with the contents like in my `CustomJsonLanguageConverter` and re-register new json converter somewhere at your application startup

**Comment by jdahan91** [comment]
@rooonikz  I have created the class with the overridden method, where do I modify it to be used?

**Comment by rooonikz** [comment]
> @rooonikz I have created the class with the overridden method, where do I modify it to be used?

What type of app do you have? Console app, ASP.NET app, ASP.NET Core app ?

**Comment by jdahan91** [comment]
Applying the changes, i'm still getting null from UserAsync.GetAuthenticatedUser().

**Comment by GodsWithin** [comment]
@rooonikz @jdahan91 I mean he is making a special converter for it, why not just change the current language converter to return `Language.English` when `reader.Value` is equal to `null` like so:

```cs
public override object ReadJson(JsonReader reader, Type objectType, object existingValue, JsonSerializer serializer)
{
      if (reader.Value == null)
      {
            return Language.English;
      }

      int parsed;

      if (int.TryParse(reader.Value.ToString(), out parsed))
      {
            return LanguageExtension.GetLangFromDescription(parsed);
      }

      return LanguageExtension.GetLangFromDescription((string)reader.Value);
}
```

In this file https://github.com/linvi/tweetinvi/blob/master/Tweetinvi.Logic/JsonConverters/JsonLanguageConverter.cs

**Comment by jdahan91** [comment]
@GodsWithin I've done that, and still get a null from the api.

**Comment by GodsWithin** [comment]
@jdahan91 Can you be more specific "null from the API", do you mean that the library is still throwing a NullReferenceException? And if so can you provide the StackTrace.

**Comment by rooonikz** [comment]
> @GodsWithin I've done that, and still get a null from the api.

Did you register new converter in `JsonPropertyConverterRepository` ?

**Comment by GodsWithin** [comment]
> > @GodsWithin I've done that, and still get a null from the api.
> 
> Did you register new converter in `JsonPropertyConverterRepository` ?

This is just an adjustment of the already existing `JsonLanguageConverter `and thus is already registered in the `JsonPropertyConverterRepository`

**Comment by jdahan91** [comment]
@GodsWithin @rooonikz Anyone of you willing to help me over Skype?

**Comment by rooonikz** [comment]
> > > @GodsWithin I've done that, and still get a null from the api.
> > 
> > 
> > Did you register new converter in `JsonPropertyConverterRepository` ?
> 
> This is just an adjustment of the already existing `JsonLanguageConverter `and thus is already registered in the `JsonPropertyConverterRepository`

This was reply for @jdahan91 :) I'm talking about the case, when you don't wan't (can't) rebuild original assembly and here you can register you own overrided class

**Comment by jdahan91** [comment]
@rooonikz I made the change in the existing JsonLanguageConverter class, just as above. I rebuilt the Logic module and replaced it inside the packages dir but still returning null.

**Comment by GodsWithin** [comment]
> @GodsWithin @rooonikz Anyone of you willing to help me over Skype?

Don't have Skype but you can add me on Discord (Gods Within#3550)

**Comment by rooonikz** [comment]
> @rooonikz I made the change in the existing `JsonLanguageConverter` class, just as above. I rebuilt the Logic module and replaced it inside the packages dir but still returning null.

I got you. I am not using that approach so, probably, @GodsWithin can help you with it

**Comment by jdahan91** [comment]
@GodsWithin Do you have Zoom?

**Comment by GodsWithin** [comment]
> @GodsWithin Do you have Zoom?

No, I don't even know what that is. I'll only be able to chat through Discord.

**Comment by jdahan91** [comment]
@GodsWithin It's amazing that I have opened an account almost on all VoIP platforms just to chat with individuals. Now I have to open another one at Discord just to talk to you lol.

**Comment by jdahan91** [comment]
@GodsWithin Can you give me your email and we can chat via Hangouts?

**Comment by GodsWithin** [comment]
I'm afraid I can't. Discord is all I can do.

**Comment by jdahan91** [comment]
@GodsWithin Then lets continue chatting here. Ok look, what I did is modified the ReadJson method inside JsonLanguageConverter class. I've rebuilt the module and copied it to the netstandard1.4 directory inside .nuget folder of tweetinviapi. It worked for me yesterday but now it's not, what's going on here?

**Comment by linvi** [comment]
Quick comment guys, there is a refactoring that I already performed for the problem of async.

You can find it here : https://github.com/linvi/tweetinvi/commits/AsyncLocalRefactoring.

As I explained to some of you, I do have very limited amount of time to dedicate on the issue but 5.0 will be fixing all of these.

**Comment by GodsWithin** [comment]
@jdahan91 I'd have to see the error you're getting. Just dump the stracktrace here or on pastebin.

**Comment by jdahan91** [comment]
@GodsWithin There's no stack trace because the exception is thrown from Tweetinvi (external code)

**Comment by GodsWithin** [comment]
If you are using Visual Studio and your code, a popup will appear where the Exception is thrown from. There should be a "View Details" button which opens another screen where the stacktrace *should* be.

**Comment by jdahan91** [comment]
I'm trying to explain to you there's no "View Details" because it's an exception from external code, the only thing the stack trace shows is `[External code]`

**Comment by GodsWithin** [comment]
Then what you should do is clone the Tweetinvi repository, open it in Visual Studio, add your code to it and try to replicate the problem.

**Comment by jdahan91** [comment]
@linvi Using the mentioned branch, I'm getting errors in my code.

`PublishTweetOptionalParameters` could not be found or missing.
`PublishTweetParameters` does not take 2 parameters.

**Comment by jdahan91** [comment]
@linvi Using the mentioned branch, i'm getting 

> Unhandled Exception:
> 
> Tweetinvi.Exceptions.TwitterException: https://api.twitter.com/1.1/account/verify_credentials.json?skip_status=true&include_entities=true&include_email=true web request failed.

**Comment by jdahan91** [comment]
@GodsWithin I did that, and i'm receiving a lot of Newtonsoft.Json error (reference not found) in all Tweetinvi projects, which I did install it from nuget manager.

**Comment by GodsWithin** [comment]
Can you show me the error, copy paste the error, or at least a screenshot. I can't really help you otherwise.

**Comment by jdahan91** [comment]
Bahh... I give up, I'm adding you on Discord.

**Comment by GodsWithin** [comment]
@jdahan91 Scroll down in discord lol

**Comment by infoskirmish** [comment]
I can confirm that the hotfix listed above works... be sure you are replacing the DLLs located here:
C:\Users\[username]\.nuget\packages\tweetinviapi\4.0.0\lib\netstandard1.4

Rebuild your app which uses these DLLs and you should be fine.

**Comment by Scobiform** [comment]
Thanks for the heads up guys ♥

**Comment by intergleam** [comment]
any ATA for fixing this? thanks

**Comment by matt123miller** [comment]
Has anybody successfully built the tweetinvi.logic.dll for .NET 4.6 or similar targets?

**Comment by matt123miller** [comment]
For now I've created my own class identical to the one hotfixed in https://github.com/linvi/tweetinvi/pull/852 and replaced the existing broken class with this one in the `Tweetinvi.TweetinviContainer`  dependecy container

**Comment by SreenivasReddyThotli** [comment]
> 
> 
> Has anybody successfully built the tweetinvi.logic.dll for .NET 4.6 or similar targets?



> 
> 
> In my case, I downloaded the ZIP file and just saved it to my hard drive. Opened the Tweetinvi.sln file in Visual Studio, made the correction to the file/line that @GodsWithin suggested ([#850 (comment)](https://github.com/linvi/tweetinvi/issues/850#issuecomment-494501533)) and then built the solution.
> 
> In my main solution I use TweetInvi in, I removed the NuGet version of TweetInvi and then added the DLL files from the TweetInvi solution I just finished building. Test and release. Worked great.
> 
> When something works so well for so long without a hiccup, we get spoiled. Thanks to everyone for the quick reaction and suggestions.

Can you please share the Tweetinvi.dll after fixing the issue.

**Comment by jdahan91** [comment]
Guys, I just want to mention that it's not enough to bypass the exception by checking for null. There's beyond that, at the core of the library that is not in sync with Twitter's api.

**Comment by SachinUnfyd** [comment]
I'm using a very old file in my project.
At that time, I did not know which solution followed the path.(https://github.com/linvi/tweetinvi )

I am currently facing some an errors.
So if I replace the tweetinvi dll file then, I'm facing the error given by the screenshot.

Please help if there is some solution on this

**Comment by SachinUnfyd** [comment]
I'm using a very old file in my project.
At that time, I did not know which solution followed the path.(https://github.com/linvi/tweetinvi )

I am currently facing some an errors.
So if I replace the tweetinvi dll file then, I'm facing the error given by the screenshot.

Please help if there is some solution on this
![ReplaceTweetinvi](https://user-images.githubusercontent.com/50991115/58370080-f4d7ae00-7f1f-11e9-8657-27cc1ed40de2.png)

**Comment by mitchcapper** [comment]
Yeah  PR #863 talks about this issue

**Comment by DavidBusayo** [comment]
> > > https://github.com/linvi/tweetinvi/blob/master/Tweetinvi.Logic/JsonConverters/JsonLanguageConverter.cs#L14 on this line, the reader.Value is null, and when you call reader.Value.ToString() it'll throw the NullReferenceException, which causes everything to break.
> > > So it seems that it has to do with the Json.NET library.
> > 
> > 
> > If you add a null check before trying to parse, it'll not throw the exception and from there no other exceptions will be thrown and the GetAuthenticatedUser() method will return properly.
> 
> I've just applied this fix and it works again
how to apply hot fix and where it should be apply. i have only a .dll file. Kindly help me out!

**Comment by DavidBusayo** [comment]
where to apply hot fix? can you guys upload that hotfix so that many beginners like me can use that modified dll.

**Comment by rooonikz** [comment]
> where to apply hot fix? can you guys upload that hotfix so that many beginners like me can use that modified dll.
You can avoid replacing dll by using my workaround in this comment. [https://github.com/linvi/tweetinvi/issues/850#issuecomment-494746515](there)

**Comment by dynamicons** [comment]
Dear roonikz, everyone,

since quite a few people seem to have a problem compiling that TweetInvi.Logic.dll with the bugfix described, here it is:

http://dynamic-apps.org/downloads/custom/

in that directory you'll find my version of the method in JsonLanguageConverter.cs. Since it's the only method in that file, you can replace it and recompile the DLL yourself. Also i have uploaded the DLL itself in the hope that it doesn't violate any license (otherwise please just tell me, don't sue me). But this is an open source project anyways, right? - wouldn't know how to integrate my solution with Github.

I have given that bugfixed Tweetinvi.Logic.dll the version number 4.0.0.1. 
use everything on your own risk of course.  

Best Regards, Martin Bernhardt.

**Comment by DavidBusayo** [comment]
@dynamicons i have modified **TweetInvi.Logic.dll** and build the solution. after that i copies all dll files from Tweetinvi bin folder and added into my project. before adding i have deleted nuget package of tweetinvi. 
still facing error and exceptions.

**Comment by DavidBusayo** [comment]
![image](https://user-images.githubusercontent.com/51048612/58384773-718d8980-7fff-11e9-961d-3fb1e76f78d5.png)

**Comment by jdahan91** [comment]
@dynamicons After pasting the modified .dll from your link to the tweetinviapi folder i'm still getting (null) as a return value from the apis.

**Comment by mitchcapper** [comment]
The better solution is as in the PR to mark the types as nullable rather than modifying the deserializer

**Comment by NagarajuRaghava** [comment]
[This](https://github.com/linvi/tweetinvi/issues/850#issuecomment-494746515) comment solved the issue for me. 
Mine is a ASP.NET Core app, added the JsonConverters remove & add lines in Startup.cs, inside `Configure` method and CustomJsonLanguageConverter class somewhere is the app. Working fine now.

**Comment by DavidBusayo** [comment]
@NagarajuRaghava  in which **class** put these lines?
JsonPropertyConverterRepository.JsonConverters.Remove(typeof(Language));
JsonPropertyConverterRepository.JsonConverters.Add(typeof(Language), new CustomJsonLanguageConverter());

**Comment by DavidBusayo** [comment]
> > where to apply hot fix? can you guys upload that hotfix so that many beginners like me can use that modified dll.
> 
> @DavidBusayo You can avoid replacing dll by using my workaround in this comment. [there](https://github.com/linvi/tweetinvi/issues/850#issuecomment-494746515)

where to put these lines?
```
JsonPropertyConverterRepository.JsonConverters.Remove(typeof(Language));
JsonPropertyConverterRepository.JsonConverters.Add(typeof(Language), new CustomJsonLanguageConverter()); // your brand new json converter with workaround
```

**Comment by NagarajuRaghava** [comment]
> where to put these lines?
> 
> ```
> JsonPropertyConverterRepository.JsonConverters.Remove(typeof(Language));
> JsonPropertyConverterRepository.JsonConverters.Add(typeof(Language), new CustomJsonLanguageConverter()); // your brand new json converter with workaround
> ```

You basically need to add these two lines in your application startup. 
If yours is a ASP.NET CORE app, you can add it inside the `Configure` method in Startup.cs class.

**Comment by dynamicons** [comment]
Dear David,

i solved it so that i simply overwrite the DLL Tweetinvi.Logic.dll in the Nuget packages subdirectory of my project. So i have replaced that one DLL within these two folders:

C:\Users\martin\Documents\Projects\StartupInformer\packages\TweetinviAPI.4.0.0\lib\net461
C:\Users\martin\Documents\Projects\StartupInformer\packages\TweetinviAPI.4.0.0\lib\netstandard1.4

Then i recompile my project and it worked for me, on the spot. 

In theory though, exchanging the linked DLL TweetInvi.Logic.dll that you find in the project's list of libraries to another location should also work. Needless to say, keep all other DLLs in place as they are.

I have to add though that all i am using from TweetInvi is LoginUser(), PublishTweet() and PublishRetweet(), and looking up an authentified user's Timeline of own postings (Tweets).

So if you are using additional functionality, keep in mind that Twitter has deprecated quite a lot of features, so there may be more issues. For a proper solution, i am afraid we all have to wait for @linvi.

Let me know if you got questions. I can show you how i succeeded on Skype.
My Skype id is:   martin.bernhardt.2002

In case you can't get it to work, i'd recommend to clone the TweetInvi Sources here from Github, which is very easy with Visual Studio. Then you simply make a directory copy of that clone, double-click the .sln file, and exchange the Sourcecode file i put on FTP and recompile.

Here again the path in case that anyone else wants to work on the final solution (Sourcecode, DLL).

http://dynamic-apps.org/downloads/custom/

Best Regards, Martin Bernhardt
www.dynamic-applications.com

**Comment by jdahan91** [comment]
@dynamicons Dude, I replaced the .dll in both folders, rebuilt but still get null from tweetinvi apis. Please help me with this, it worked for me when I did this the first time but now it stopped working.

**Comment by juanjhb1894** [comment]
> Dear David,
> 
> i solved it so that i simply overwrite the DLL Tweetinvi.Logic.dll in the Nuget packages subdirectory of my project. So i have replaced that one DLL within these two folders:
> 
> C:\Users\martin\Documents\Projects\StartupInformer\packages\TweetinviAPI.4.0.0\lib\net461
> C:\Users\martin\Documents\Projects\StartupInformer\packages\TweetinviAPI.4.0.0\lib\netstandard1.4
> 
> Then i recompile my project and it worked for me, on the spot.
> 
> In theory though, exchanging the linked DLL TweetInvi.Logic.dll that you find in the project's list of libraries to another location should also work. Needless to say, keep all other DLLs in place as they are.
> 
> I have to add though that all i am using from TweetInvi is LoginUser(), PublishTweet() and PublishRetweet(), and looking up an authentified user's Timeline of own postings (Tweets).
> 
> So if you are using additional functionality, keep in mind that Twitter has deprecated quite a lot of features, so there may be more issues. For a proper solution, i am afraid we all have to wait for @linvi.
> 
> Let me know if you got questions. I can show you how i succeeded on Skype.
> My Skype id is: martin.bernhardt.2002
> 
> In case you can't get it to work, i'd recommend to clone the TweetInvi Sources here from Github, which is very easy with Visual Studio. Then you simply make a directory copy of that clone, double-click the .sln file, and exchange the Sourcecode file i put on FTP and recompile.
> 
> Here again the path in case that anyone else wants to work on the final solution (Sourcecode, DLL).
> 
> http://dynamic-apps.org/downloads/custom/
> 
> Best Regards, Martin Bernhardt
> [www.dynamic-applications.com](http://www.dynamic-applications.com)

I did all the steps and it works on the first time with an specific credentials, but when I tried the PIN-Based Authentication it stop working

**Comment by demirag** [comment]
Hi Guys, Thanks for the info. As everyone else I am also affected from the twitter change. Do you know who can release a new version with these fixes? This is really an emergency for everyone.

**Comment by linvi** [comment]
I will be attempting to release a fix tomorrow. Sorry for the long time to reply.

Cheers,
Linvi

**Comment by demirag** [comment]
Thank you @linvi . Can't wait for the release.

**Comment by linvi** [comment]
The new version of Tweetinvi is now available. Please use nuget package manager to update to 4.0.1.

``` sh
Install-Package TweetinviAPI -Version 4.0.1
```

**Comment by SachinUnfyd** [comment]
hi..
I've been update DLL.
I'm having this error

![tweetBug](https://user-images.githubusercontent.com/50991115/58451784-97677b00-8132-11e9-9e78-83414f88c573.png)

**Comment by SachinUnfyd** [comment]
I've been update dLL.
I have this error
![tweetBug2](https://user-images.githubusercontent.com/50991115/58452068-a00c8100-8133-11e9-9f38-27f8afef113f.png)

**Comment by SreenivasReddyThotli** [comment]
Please let us know once it's fixed.  We are waiting for your reply

**Comment by jdahan91** [comment]
I'm getting null from `AuthFlow.InitAuthentication()`.

**Comment by jamiemaguiredotnet** [comment]
Yeah, the API is broken right now, I left a comment on the Twitter Dev forums too.

**Comment by jdahan91** [comment]
OMG.... All this time, the problem was that the DateTime on my android emulator was not synced. I closed and reopened the emulator again and now everything is back to normal.... Jesus...

**Comment by matt123miller** [comment]
I'm experiencing issues with System.Net.Http versions. It originally complained about 4.0.0 and I've been trying to follow some stack overflow suggestions and got this far. I'm targetting .net framework 4.6.1 if that's relevant.

![image](https://user-images.githubusercontent.com/9091121/58466432-941fbf80-8131-11e9-9fca-f3cb62becbbc.png)

**Comment by jamiemaguiredotnet** [comment]
I d/l the source and started to debug, someone on Twitter dev coded a workaround (didnt work for me)


The more I debugged, it seems there is a problem with the Language property.


________________________________

**Comment by sirkris** [comment]
I was having the same issue as reported by OP and others.  Specifically, Tweetinvi.User.GetUserFromScreenName was returning null in my code as a result.

**I updated the TweetinviAPI dependency in NuGet to 4.0.1 and that fixed it**, in case anybody's wondering.  In the package manager console, simply entering "Update-Package TweetinviAPI" should do the trick.

**Comment by linvi** [comment]
Please all can you please thumbs up/down if you are still experiencing issues after 4.0.1.

If you do and you can execute `User.GetAuthenticatedUser()` please open another issue.

**Comment by linvi** [comment]
Hi all can you please thumbs up/down if you are still experiencing issues after 4.0.1.

If you do and you can execute `User.GetAuthenticatedUser()` please open another issue.

I also wanted to thank you all for opening CRs and helping resolve the problem.

@sirkris @jamiemaguiredotnet @matt123miller @SamusAranX @ashr @JAlbrightDev @tsconn23 @jdahan91 @GodsWithin @DeschaumesDaan @jcmag @z0o0x @soatley @DavidBusayo @NagarajuRaghava @dynamicons @juanjhb1894 @demirag @SachinUnfyd @SreenivasReddyThotli @pronet99 @sameersb @chethi666 @intergleam @gribunin @AlphaNETRD @chenKerman @jxlarrea @dontbesorry80

**Comment by snerpton** [comment]
> The new version of Tweetinvi is now available. Please use nuget package manager to update to 4.0.1.
> 
> ```shell
> Install-Package TweetinviAPI -Version 4.0.1
> ```
> 
> I will be closing the issue tomorrow if nobody has any compatibility issues.

Perfect, been working for me for the past hour or two. Thanks for the quick fix.

**Comment by NullReferenceEric** [comment]
I'm getting Could not load file or assembly 'System.Runtime' or one of its dependencies. The located assembly's manifest definition does not match the assembly reference. (Exception from HRESULT: 0x80131040) when I update to 4.0.1.  When I roll back to 4.0.0 works fine.  Any ideas?

**Comment by linvi** [comment]
It looks like a problem of reference in your `app.config`. 

It should be something similar to : https://github.com/linvi/tweetinvi/wiki/System.Net.Http-not-found-exception

**Comment by NullReferenceEric** [comment]
> @NullReferenceEric
> 
> It looks like a problem of reference in your `app.config`.
> 
> It should be something similar to : https://github.com/linvi/tweetinvi/wiki/System.Net.Http-not-found-exception

Thanks for the help.  I have the line below in my web.config.  Did anything change in  4.0.1?  If I rollback to 4.0.0 this isn't an issue.

` <dependentAssembly>
        <assemblyIdentity name="System.Net.Http" publicKeyToken="b03f5f7f11d50a3a" culture="neutral" />
        <bindingRedirect oldVersion="0.0.0.0-4.3.4.0" newVersion="4.0.0.0" />
      </dependentAssembly>`

**Comment by demirag** [comment]
I have just upgraded and receiving similar error as @NullReferenceEric.  

Could not load file or assembly 'System.Runtime, Version=4.0.20.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. The system cannot find the file specified.

No matter what I do on web.config bindings, I could not solve the problem.

**Comment by linvi** [comment]
@NullReferenceEric @demirag could one of you share your project or a sample for which you can reproduce the issue?

**Comment by mr3chitk** [comment]
i temporarily solved the runtime 4.2 problem by unref webhooks.dll in csproj

**Comment by muratyagis** [comment]
by the way my project is not an ASP.NET, is a Windows Form project

**Comment by dorislinda** [comment]
@linvi  I try to install package TweetinviAPI -Version 4.0.1 in Visual Studio 2015 but failed  it seems NitoAsyncEx.Context.StrongName 1.1.0 stops the installing

**Comment by jdahan91** [comment]
> i temporarily solved the runtime 4.2 problem by removing webhooks.dll ref in csproj

I did the same thing lol, just removed the .dll from the nuget package folder.

**Comment by matt123miller** [comment]
> I'm experiencing issues with System.Net.Http versions.
I managed to fix my issue, ended up experiencing the `system.runtime 4.2.0.0` issue along the way. Finally installed and tried DotPeak from Jetbrains. Really helped identify what DLLs were which version etc.

I manually added a reference to system.runtime 4.3.1 via nuget then added the following binding redirect:
``` 
<dependentAssembly> 
    <assemblyIdentity name="System.Runtime" publicKeyToken="b03f5f7f11d50a3a" culture="neutral" /> 
    <bindingRedirect oldVersion="0.0.0.0-4.3.1.0" newVersion="4.1.2.0" /> 
</dependentAssembly> 
```
This seems to work in my use case, glad to get a fix pushed out to users. Thanks @linvi and everyone involved!

**Comment by AbePerl1** [comment]
I still have an issue with the latest version 4.01 as well.

Could not load file or assembly 'System.Collections, Version=4.1.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. The system cannot find the file specified.

**Comment by peterpociask** [comment]
Is it possible to apply this fix to earlier versions? I have an app using 1.1 that cannot be updated.

**Comment by NullReferenceEric** [comment]
> Is it possible to apply this fix to earlier versions? I have an app using 1.1 that cannot be updated.

Yes.  You can download the source solution and apply the fix mentioned above, recompile and use the dlls.  Here's the comment

https://github.com/linvi/tweetinvi/issues/850#issuecomment-495205035

**Comment by cblaze22** [comment]
This System.Runtime error is killing me here. Trying everything and not sure why it is appearing on this .1 update.

**Comment by AbePerl1** [comment]
> I still have an issue with the latest version 4.01 as well.
> 
> Could not load file or assembly 'System.Collections, Version=4.1.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. The system cannot find the file specified.

Now it works for me. I re-created the project in .netCore

**Comment by monty241** [comment]
Same issue as @AbePerl1. Project in Framework 4.7.2, can not switch to .net core since it is really hard bound to framework. Reverted to 3.0.0 and added for now code to ignore the NullReferenceException. The tweet gets sent before the NullReferenceException is raised.

**Comment by dberroa** [comment]
I'm on .net framework 4.6.2 and was working fine compile wise with 4.0.0 but 4.0.1 is now erroring 

`Assembly 'PROJECT' with identity 'PROJECT, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null' uses 'System.Net.Http, Version=4.2.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' which has a higher version than referenced assembly 'System.Net.Http' with identity 'System.Net.Http, Version=4.1.1.3, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'	PROJECT.Tests`

I'm not sure why this is happening.

**Comment by cblaze22** [comment]
I somehow finally got it fixed, dont ask me how.  Whoever made this update did more then fix that one piece of code.

**Comment by dberroa** [comment]
I still can't get it to work. With no changes to the rest of my application, I'm now getting higher referenced version errors like the one I posted above.

**Comment by WhiteLionATX** [comment]
While 4.0.1 throws null ref when getting credentials ("Tweetinvi.Auth"). The v 4.0.0 throws NullReferenceException in "Newtonsoft.Json.dll". I m still waiting for a solution. Any news on this ?

**Comment by nielsodgaard** [comment]
I'm having the same issue using version 1.3.0. If the issue has been fixed in version 4.0.1, we probably need the same fix applied to the other major versions aswell.

**Comment by alanwthom** [comment]
A late reply, but for anyone who is having assembly issues after updating to `4.0.1` this might be helpful (I'm targeting NetFramework461). It seems like there's an issue with the assembly references on NuGet install, so I got around this by doing a "manual" reference update. Hope this can help someone else.

1. Update the version of TweetinviAPI in your `packages.config` file to 4.0.1
2. Run a NuGet restore to install version `4.0.1`
3. Remove all references to TweetinviAPI from your project (Your Project > References > Tweetinvi > right-click > Remove - There will be about 9 and all should be the "old" version if you check their properties)
4. Add references to all of the "new" Tweetinvi dll's (Your Project > References > right-click > Add Reference > Browse > Packages > TweetinviAPI.4.0.1 > lib > 461 - You should get the same 9 dll's> Add).
5. Rebuild your solution

**Comment by dberroa** [comment]
@alanwthom 

I tried your method and while I did see the assembly number increase to 4.0.1 in the properties of the DLL I'm still having the same referencing issue between my projects:

Assembly '****Core' with identity '****Core, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null' uses 'System.Net.Http, Version=4.2.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' which has a higher version than referenced assembly 'System.Net.Http' with identity 'System.Net.Http, Version=4.1.1.3, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'

**Comment by mphacker** [comment]
Still having reference issues in .NET Framework apps.  Is this actively being worked on and something that will be resolved soon or is this a bigger problem?

**Comment by linvi** [comment]
@mphacker please check https://github.com/linvi/tweetinvi/issues/871

I have not managed to find a solution for that problem just yet but for me using .NET 4.8 solved the problem.

If you could please comment/share a repro git it would be great.
Please comment on the other issue.

**Comment by d668** [comment]
So there is still no fix? basically the repo is unusable?

**Comment by SachinUnfyd** [comment]
i am facing Same issue,

var autuser = Tweetinvi.User.GetAuthenticatedUser();
var user = Tweetinvi.User.GetUserFromScreenName("@*********");

autuser or user get always null response

**Comment by sirkris** [comment]
Try removing the @ char.

**Comment by scgough** [comment]
I've got the runtime issue in a project using .NET 4.6.1 / MVC5 (VS 2015) 
Is upgrading to 4.8 the only way around this at the moment?

**Comment by linvi** [comment]
Hello,

Sorry for the long response, I am back from holidays and will be getting back to the issue of the application not properly working with .NET Framewrok 4.7.2.

If anyone faced the same problem, please feel free to share your knowledge as last time I checked I was blocked by an open bug on Microsoft Framework.

Cheers,
Linvi

**Comment by linvi** [comment]
I have updated the version of Tweetinvi to 4.0.2 so that it becomes compatible with .NET Framework 4.7.2.

Please note that the it came with a cost which is the fact that for the moment Tweetinvi is no longer a StrongName package.

I will be glad to hear feedback on the subject before I close the issue.

**Comment by linvi** [comment]
@dberroa did you try to update System.Net.Http nuget package to the latest version?

**Comment by linvi** [comment]
Have you tried first removing Tweetinvi and then System.Net.Http and then install Tweetinvi.

**Comment by linvi** [comment]
@dberroa if you are working on it right please feel free to contact me on https://gitter.im/linvi

**Comment by linvi** [comment]
As no additional feedback, I am now closing this issue.

Thank you all for participating and reporting the issue.

**Comment by MohamadMousheimish** [comment]
So what was the solution for this issue?? I'm facing it now

**Comment by linvi** [comment]
You need to update and use latest version of Tweetinvi.

**Comment by Johnt201** [comment]
I reinstalled Tweetinviapi on VS 2017 three days ago.  I didn't check the version then but today it is 4.0.3 and webhooks 4.0.0.  I get the Null value with: var TwitResult = Tweetinvi.User.GetAuthenticatedUser();

Also, when I try the "Hello World" publish example it does not report a problem but there is no publish on my timeline.  
I checked and double checked the keys and secrets, and ever regenerated new ones and copy and pasted them.

Here's my code: 
        //   Auth.SetUserCredentials(ConsumerKey, ConsumerSecret, AccessToken, TokenSecret);
        try
        {
            Tweet.PublishTweet("Hello World");  //nothing showed up on my timeline
            Twit.Text = "Published tweet";
        }
        catch {
            Twit.Text = "Try failed";
        }


        var TwitResult = Tweetinvi.User.GetAuthenticatedUser();
        if (TwitResult == null)
        {
            NL.Text = "Null value";  //this is where I'm stuck
            Label1.Text = ExceptionHandler.GetLastException().TwitterDescription;
        }
        else
        {
            Label1.Text = TwitResult.ToString();
        }

**Comment by linvi** [comment]
@John201 It does look like another issue. Could you please open a new ticket if you still encounter this problem.

**Comment by linvi** [comment]
I am locking this ticket which has become way too big now, please feel free to open new tickets and link to that one if you need further help.
